File: prog2/Special/Define.cs
// Define -- Parse tree node strategy for printing the special form define

using System;

namespace Tree
{
    public class Define : Special
    {
	public Define() { }

        public override void print(Node t, int n, bool p)
        {
            Printer.printDefine(t, n, p);
        }

        public override Node eval(Node t, Environment env)
        {
          Node iden = t.getCdr().getCar();
          Node val = t.getCdr().getCdr().getCar();

          if (iden.isSymbol()) 
          {
            env.define(iden, val);
          }
          else 
          {
            Closure close = new Closure(new Cons(t.getCdr().getCar().getCdr(), t.getCdr().getCdr()), env);
            env.define(iden.getCar(), close);
          }
          if(iden.isPair()) {
            Node func = iden.getCar();

            Cons list = new Cons(new Ident("lambda"), new Cons(val, t.getCdr().getCar().getCdr()));
            env.define(func, list);
          }
          return new StringLit("Values are null (Define)");
        }
    }
}

File: prog1/Define.java
class Define extends Special {
  void print(Node t, int n, boolean p, boolean q) {
    if(q) System.out.println();
    System.out.print(" ".repeat(n));

    if (!p) {
      System.out.print("(");
    }
    t.getCar().print(0);

    int minNumSpaces = n > 1 ? n : 1;
    int indentSpace = t.getCdr().isNull() ? 0 : minNumSpaces;
    t.getCdr().print(indentSpace, true, q);
  }
}
